### PR TITLE
Simplify conditional catalyst expressions generated by udf-compiler

### LINKED ISSUE #3985: [QST] UDF-Compiler: Predicate pushdown is not working in this case
**What is your question?**
Hi, It is a great open-source project, thanks for sharing. We have been evaluating the udf-compiler, it works great as advertised, but we run into a problem when we try to test a filter pushdown. We wrote a simple UDF as below, but the PushedFilters is empty with udf-compiler:
```
    spark.udf.register("isOlderThan20", (i: Int) => {i>20})
    val udfResult = spark.sql("SELECT * FROM people_with_schema WHERE isOlderThan20(age)")
    
   +- FileScan json [name#0,age#1] Batched: false, DataFilters: [if ((age#1 > 20)) true else false], ..... PushedFilters: [], 
```
What we expected is to have a physical plan like this:
```
    val udfResult = spark.sql("SELECT * FROM people_with_schema WHERE age > 20")
   
    !+- Filter (isnotnull(age#1) AND (age#1 > 20))                                                                                                                                                                                                                                                     
    +- FileScan ...PushedFilters: [IsNotNull(age), GreaterThan(age,20)], 
```
By turning on trace, it looks like that the udf-compiler translated this udf as "if-else", instead of just "age#1 >20":
`Filter if ((age#1 > 20)) true else false`

Wonder if this is by design? If not, wonder if you can suggest some ideas to us on how to fix this?

Thank you very much!
-Yong Wang

### PR DESCRIPTION
**Pull request by seanprime7**
if (c) true else false => c
if (c) false else true => !c

Signed-off-by: Sean Lee <selee@nvidia.com>

Resolves #3985

### CODE REVIEW

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala [hunk 1]
```diff
@@ -255,6 +255,7 @@ object CatalystExpressionBuilder extends Logging {
         case And(Literal.TrueLiteral, c) => simplifyExpr(c)
```
Comment (jlowe) [inline udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala:255]: 2022 copyrights

Comment (seanprime7) [reply udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala:255]: Updated the copyright years

**Review by jlowe — DISMISSED** [review]

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala [hunk 2]
```diff
@@ -255,6 +255,7 @@ object CatalystExpressionBuilder extends Logging {
...
         case And(c, Literal.TrueLiteral) => simplifyExpr(c)
         case And(Literal.FalseLiteral, _) => Literal.FalseLiteral
+        case And(_, Literal.FalseLiteral) => Literal.FalseLiteral
```
Comment (gerashegalov) [inline udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala:258]: Can be collapsed to a single line
```suggestion
        case and: And if and.containsChild(Literal.FalseLiteral) => Literal.FalseLiteral
```

**Review by abellina — DISMISSED** [review]
LGTM other than @gerashegalov's suggestions.

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala (hunk 2 above)
Comment (seanprime7) [reply udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala]: done

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala [hunk 3]
```diff
@@ -347,6 +348,7 @@ object CatalystExpressionBuilder extends Logging {
         }
         case And(c1, c2) => And(simplifyExpr(c1), simplifyExpr(c2))
         case Or(Literal.TrueLiteral, _) => Literal.TrueLiteral
+        case Or(_, Literal.TrueLiteral) => Literal.TrueLiteral
```
[the remark this answers is not in the document]
Comment (seanprime7) [reply udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala]: done

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by jlowe** [comment]
@seanprime7 please do not force-push once the PR is being reviewed.  It often orphans the reviewer comments and makes re-reviews more difficult.  Commits are already squashed when merged, so PRs don't need to optimize their commits before merging.

**Comment by seanprime7** [comment]
copy that

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Pull request merged**